src/jit/utils.cpp

```diff
@@ -1954,6 +1954,68 @@ float FloatingPointUtils::round(float x)
     return _copysignf(flrTempVal, x);
 }
 
+bool FloatingPointUtils::isNormal(double x)
+{
+    int64_t bits = *(int64_t*)&x;
+    bits &= 0x7FFFFFFFFFFFFFFF;
+    return (bits < 0x7FF0000000000000) && (bits != 0) && ((bits & 0x7FF0000000000000) != 0);
+}
+
+bool FloatingPointUtils::isNormal(float x)
+{
+    int32_t bits = *(int32_t*)&x;
+    bits &= 0x7FFFFFFF;
+    return (bits < 0x7F800000) && (bits != 0) && ((bits & 0x7F800000) != 0);
+}
+
+//------------------------------------------------------------------------
+// hasPreciseReciprocal: check double for precise reciprocal. E.g. 2.0 <--> 0.5
+//
+// Arguments:
+//    x - value to check for precise reciprocal
+//
+// Return Value:
+//    True if 'x' is a power of two value and is not denormal (denormals may not be well-defined
+//    on some platforms such as if the user modified the floating-point environment via a P/Invoke)
+//
+
+bool FloatingPointUtils::hasPreciseReciprocal(double x)
+{
+    if (!isNormal(x))
+    {
+        return false;
+    }
+
+    uint64_t i        = *(uint64_t*)&x;
+    uint64_t exponent = (i >> 52) && 0x7FFul;
+    uint64_t mantissa = i & 0xFFFFFFFFFFFFFul;
+    return mantissa == 0 && exponent != 0 && exponent != 1023;
+}
+
+//------------------------------------------------------------------------
+// hasPreciseReciprocal: check float for precise reciprocal. E.g. 2.0f <--> 0.5f
+//
+// Arguments:
+//    x - value to check for precise reciprocal
+//
+// Return Value:
+//    True if 'x' is a power of two value and is not denormal (denormals may not be well-defined
+//    on some platforms such as if the user modified the floating-point environment via a P/Invoke)
+//
+
+bool FloatingPointUtils::hasPreciseReciprocal(float x)
+{
+    if (!isNormal(x))
+    {
+        return false;
+    }
+
+    uint32_t i        = *(uint32_t*)&x;
+    uint8_t  exponent = (uint8_t)(i >> 23);
```

Comment:
Why the cast to `uint8_t`?

Reply:
@mikedn oh, it was a not really obvious way to drop the sign bit. Will replace the cast with `& 0xFF`

Reply:
> oh, it was a not really obvious way to drop the sign bit.

Exactly :smile:. Not only that it isn't obvious but some analyzers may complain about such truncation (I'm pretty sure VC++'s `/RTCc` would)